flink-runtime/src/main/java/org/apache/flink/runtime/deployment/ConsumedSubpartitionContext.java

```diff
@@ -0,0 +1,144 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.deployment;
+
+import org.apache.flink.runtime.executiongraph.IndexRange;
+import org.apache.flink.runtime.executiongraph.IntermediateResult;
+import org.apache.flink.runtime.executiongraph.IntermediateResultPartition;
+import org.apache.flink.runtime.jobgraph.IntermediateResultPartitionID;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/*
+ * Helper class used to track and manage the relationships between shuffle descriptors and their
+ * associated subpartitions.
+ */
+public class ConsumedSubpartitionContext implements Serializable {
+    /** The number of consumed shuffle descriptors. */
+    private final int numConsumedShuffleDescriptors;
+
+    /**
+     * A mapping between ranges of consumed shuffle descriptors and their corresponding subpartition
+     * ranges.
+     */
+    private final Map<IndexRange, IndexRange> consumedShuffleDescriptorToSubpartitionRangeMap;
+
+    ConsumedSubpartitionContext(
+            int numConsumedShuffleDescriptors,
+            Map<IndexRange, IndexRange> consumedShuffleDescriptorToSubpartitionRangeMap) {
+        this.numConsumedShuffleDescriptors = numConsumedShuffleDescriptors;
+        this.consumedShuffleDescriptorToSubpartitionRangeMap =
+                checkNotNull(consumedShuffleDescriptorToSubpartitionRangeMap);
+    }
+
+    public int getNumConsumedShuffleDescriptors() {
+        return numConsumedShuffleDescriptors;
+    }
+
+    public Collection<IndexRange> getConsumedShuffleDescriptorRanges() {
+        return consumedShuffleDescriptorToSubpartitionRangeMap.keySet();
+    }
+
+    public IndexRange getConsumedSubpartitionRange(Integer shuffleDescriptorIndex) {
+        for (Map.Entry<IndexRange, IndexRange> entry :
+                consumedShuffleDescriptorToSubpartitionRangeMap.entrySet()) {
+            IndexRange shuffleDescriptorRange = entry.getKey();
+            if (shuffleDescriptorIndex >= shuffleDescriptorRange.getStartIndex()
```

Comment:
maybe checkArgument(shuffleDescriptorIndex >= shuffleDescriptorRange.getStartIndex()
                    && shuffleDescriptorIndex <= shuffleDescriptorRange.getEndIndex(), "xxxx")

Reply:
I understand now, we cannot use the check function here because we need to iterate through the entire range to determine whether the shuffler descriptor exists.